tensorflow/core/kernels/segment_reduction_ops_gpu.cu.h

```diff
@@ -1029,6 +1053,361 @@ struct SparseSegmentGradFunctor<GPUDevice, T, Index, SegmentId> {
+                            &tmp_weights, done);
+      weights_ptr = tmp_weights.flat<Tweights>().data();
+    }
+
+    const TindicesCompact* sorted_indices_ptr = indices;
+    const Tsegmentids_internal* permuted_segment_ptr = segment_ids;
+    Tensor tmp_sorted_indices;
+    Tensor tmp_permuted_segment;
+    if (dense_output_dim0 > 1) {
+      // Sort indices and permute segments.
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          context->allocate_temp(DataTypeToEnum<TindicesCompact>::value,
+                                 TensorShape({nouter}), &tmp_sorted_indices),
+          done);
+      TindicesCompact* tmp_sorted_indices_ptr =
+          tmp_sorted_indices.flat<TindicesCompact>().data();
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          context->allocate_temp(DataTypeToEnum<Tsegmentids_internal>::value,
+                                 TensorShape({nouter}), &tmp_permuted_segment),
+          done);
+      Tsegmentids_internal* tmp_permuted_segment_ptr =
+          tmp_permuted_segment.flat<Tsegmentids_internal>().data();
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          GpuRadixSort(context, nouter,
+                       /*keys_in=*/indices,
+                       /*keys_out=*/tmp_sorted_indices_ptr,
+                       /*indices_in=*/segment_ids,
+                       /*indices_out=*/tmp_permuted_segment_ptr,
+                       /*num_bits=*/Log2Ceiling64(dense_output_dim0)),
+          done);
+      sorted_indices_ptr = tmp_sorted_indices_ptr;
+      permuted_segment_ptr = tmp_permuted_segment_ptr;
+      // The original tensors are no longer needed.
+      indices_tensor = Tensor();
+      indices = nullptr;
+      segment_ids_tensor = Tensor();
+      segment_ids = nullptr;
+    }
+
+    using CountIter = gpuprim::CountingInputIterator<Toffsets>;
+    using EdgeIndicatorIter = gpuprim::TransformInputIterator<
+        Toffsets, EdgeIndicatorFunctor<TindicesCompact>, CountIter>;
+    EdgeIndicatorIter sorted_indices_edge_indicator(
+        CountIter(0),
+        EdgeIndicatorFunctor<TindicesCompact>(sorted_indices_ptr));
+
+    Tensor tmp_sorted_indices_unique_ids;
+    OP_REQUIRES_OK_ASYNC(context,
+                         context->allocate_temp(DataTypeToEnum<Toffsets>::value,
+                                                TensorShape({nouter}),
+                                                &tmp_sorted_indices_unique_ids),
+                         done);
+    Toffsets* sorted_indices_unique_ids_ptr =
+        tmp_sorted_indices_unique_ids.flat<Toffsets>().data();
+    OP_REQUIRES_OK_ASYNC(
+        context,
+        GpuInclusivePrefixSum(context, nouter, sorted_indices_edge_indicator,
+                              sorted_indices_unique_ids_ptr),
+        done);
+
+    se::Stream* stream = context->op_device_context()->stream();
+    OP_REQUIRES_ASYNC(context, stream,
+                      errors::Internal("No GPU stream available."), done);
```

Comment:
These errors have since been deprecated, and now we're supposed to use `absl::` versions, i.e. `absl::InternalError(...)`.  For concatenating error messages, this needs to be combined with `absl::StrCat(...)`.  Here and throughout.